controllers/drain_controller.go

```diff
@@ -0,0 +1,464 @@
+/*
+Copyright 2021.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"sync"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/workqueue"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	sriovnetworkv1 "github.com/k8snetworkplumbingwg/sriov-network-operator/api/v1"
+	"github.com/k8snetworkplumbingwg/sriov-network-operator/pkg/drain"
+	constants "github.com/k8snetworkplumbingwg/sriov-network-operator/pkg/global/consts"
+	"github.com/k8snetworkplumbingwg/sriov-network-operator/pkg/global/vars"
+	"github.com/k8snetworkplumbingwg/sriov-network-operator/pkg/platforms"
+	"github.com/k8snetworkplumbingwg/sriov-network-operator/pkg/utils"
+)
+
+type DrainReconcile struct {
+	client.Client
+	Scheme         *runtime.Scheme
+	drainer        drain.DrainInterface
+	resourcePrefix string
+
+	nodesInReconcile      map[string]interface{}
+	nodesInReconcileMutex sync.Mutex
+	drainCheckMutex       sync.Mutex
+}
+
+func NewDrainReconcileController(client client.Client, Scheme *runtime.Scheme, platformHelper platforms.PlatformHelperInterface) (*DrainReconcile, error) {
+	resourcePrefix := os.Getenv("RESOURCE_PREFIX")
+	if resourcePrefix == "" {
+		return nil, fmt.Errorf("RESOURCE_PREFIX environment variable can't be empty")
+	}
+
+	drainer, err := drain.NewDrainer(resourcePrefix, platformHelper)
+	if err != nil {
+		return nil, err
+	}
+
+	return &DrainReconcile{
+		client,
+		Scheme,
+		drainer,
+		resourcePrefix,
+		map[string]interface{}{},
+		sync.Mutex{},
+		sync.Mutex{}}, nil
+}
+func (dr *DrainReconcile) TryLockNode(nodeName string) bool {
+	dr.nodesInReconcileMutex.Lock()
+	defer dr.nodesInReconcileMutex.Unlock()
+
+	_, exist := dr.nodesInReconcile[nodeName]
+	if exist {
+		return false
+	}
+
+	dr.nodesInReconcile[nodeName] = nil
+	return true
```

Comment:
You've implemented a thread safe set. Maybe use https://github.com/deckarep/golang-set instead?

Reply:
Hi thanks for pointing out this one!
if possible I would like to avoid adding more packages for something really small here